lib/kafka_ex/protocol/sync_group.ex

```diff
@@ -0,0 +1,76 @@
+defmodule KafkaEx.Protocol.SyncGroup do
+  import KafkaEx.Protocol.Common
+
+  @member_assignment_version 0
+
+  defmodule Assignment do
+    defstruct topic: nil, partitions: []
+    @type t :: %Assignment{topic: binary, partitions: [integer]}
+  end
+
+  defmodule Response do
+    defstruct error_code: nil, assignments: []
+    @type t :: %Response{error_code: atom | integer, assignments: [Assignment.t] }
+  end
+
+  @spec create_request(integer, binary, binary, integer, binary, [{binary, [Assignment.t]}]) :: binary
+  def create_request(correlation_id, client_id, group_name, generation_id, member_id, assignments) do
+    KafkaEx.Protocol.create_request(:sync_group, correlation_id, client_id) <>
+    << byte_size(group_name) :: 16-signed, group_name :: binary,
+       generation_id :: 32-signed,
+       byte_size(member_id) :: 16-signed, member_id :: binary,
+       length(assignments) :: 32-signed, group_assignment_data(assignments, "") :: binary,
+       0 :: 32-signed
+    >>
+  end
+
+  @spec parse_response(binary) :: Response.t
+  def parse_response(<< _correlation_id :: 32-signed, error_code :: 16-signed,
+                        _member_assignment_len :: 32-signed,
```

Comment:
> Background: MemberAssignment is, from the broker's perspective, opaque. This allows clients to coordinate together - they join the group, one gets made leader by the broker and gets to do the member assignments, and then the leader sends whatever data the group has agreed upon to the broker to send around. Therefore, MemberAssignment is primary an opaque bytestring to the broker, hence the need to send it length+data. Then, the "consumer" protocol as implemented in the Java client goes on to interpret the MemberAssignment structure as [{topic, [partition]}] which my code supports making it theoretically viable to mix Java and kafka_ex clients (if you ever want to do that ;-)).

I sort of think this is not controversial, the part that is controversial is the API docs has `Version` as an attribute on `MemberAssignment` but there is no documentation on what `Version` means nor is it implemented in the reference implementation, which is why I agree that leaving it as `NULL` (0) is fine for now.

> On using the length: I had most of the unpacking code done when I found out about the above. It turns out that when you know the contents of the structure, you don't need the length. The Java client has a completely different approach to (de)serialization through the generic Struct class and metadata descriptions, I don't think there's much for us to pick up.

While I agree with this, this approach is inconsistent with what we do in other places, e.g:

https://github.com/kafkaex/kafka_ex/blob/master/lib/kafka_ex/protocol/fetch.ex#L76
https://github.com/kafkaex/kafka_ex/blob/master/lib/kafka_ex/protocol/fetch.ex#L83